### PR TITLE
gh-105858: Improve AST node constructors

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Demonstration:

```pycon
>>> ast.FunctionDef.__annotations__
{'name': <class 'str'>, 'args': <class 'ast.arguments'>, 'body': list[ast.stmt], 'decorator_list': list[ast.expr], 'returns': ast.expr | None, 'type_comment': str | None, 'type_params': list[ast.type_param]}
>>> ast.FunctionDef()
<stdin>:1: DeprecationWarning: FunctionDef.__init__ missing 1 required positional argument: 'name'. This will become an error in Python 3.15.
<stdin>:1: DeprecationWarning: FunctionDef.__init__ missing 1 required positional argument: 'args'. This will become an error in Python 3.15.
<ast.FunctionDef object at 0x101959460>
>>> node = ast.FunctionDef(name="foo", args=ast.arguments())
>>> node.decorator_list
[]
>>> ast.FunctionDef(whatever="you want", name="x", args=ast.arguments())
<stdin>:1: DeprecationWarning: FunctionDef.__init__ got an unexpected keyword argument 'whatever'. Support for arbitrary keyword arguments is deprecated and will be removed in Python 3.15.
<ast.FunctionDef object at 0x1019581f0>
```

* Issue: gh-105858

### CODE REVIEW

File: Parser/asdl_c.py
Comment (JelleZijlstra) [inline Parser/asdl_c.py]:
```suggestion
                    return -1;
```

File: Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst
Comment (carljm) [inline Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst]: Perhaps "optional fields are `None` if omitted" is too obvious to mention, but it also seems odd to mention the list case and not the optional case here, since optional fields defaulting to `None` (rather than unset) is also new behavior in this PR.

File: Parser/asdl_c.py
Comment (carljm) [inline Parser/asdl_c.py:998]: I guess it is safe to assume that no other type of field will ever be added to AST nodes that would be represented as a GenericAlias?

File: Parser/asdl_c.py
Comment (carljm) [inline Parser/asdl_c.py:1056]: Why do AST types even need their own custom reduce function? Shouldn't `object.__reduce__` (which IIRC uses `__new__` to safely create an empty instance, then updates its `__dict__`) work fine for AST objects as well?

File: Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst
Comment (carljm) [inline Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst:7]: Is there any way to explore the likely impact of these deprecations in the wild?

The new behavior seems much more reasonable to me; less error-prone and more inline with the usual behavior of Python objects. But it still seems like we should weigh these clear benefits against some idea of the impact/churn on downstream code.

Has this been discussed anywhere public (e.g. on Discourse) to give interested parties a chance to speak up if they have code that would require lots of updates to adhere to these new requirements?

Comment (JelleZijlstra) [reply Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst:7]: Good point, opened https://discuss.python.org/t/proposed-change-to-ast-node-constructors/46531

File: Parser/asdl_c.py
Comment (JelleZijlstra) [reply Parser/asdl_c.py:1056]: The `__new__` call will trigger DeprecationWarnings. If we don't have our own reducer, unpickling will essentially do `node = ast.FunctionDef(); node.__dict__.update(...)`, and the first line raises warnings because we don't set the required `name` field.

File: Parser/asdl_c.py
Comment (JelleZijlstra) [reply Parser/asdl_c.py:998]: "ever' is a long time, but we're restricted to types accepted by ASDL, and list is currently the only one that would be represented by a GenericAlias. If a new type were to be added, we'd be able to update this code.

File: Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst
Comment (JelleZijlstra) [reply Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst]: Good point, adding

File: Parser/asdl_c.py
Comment (JelleZijlstra) [reply Parser/asdl_c.py:998]: It might make sense to add an assertion here, equivalent to `assert type.__origin__ is list`, but it doesn't seem like we have an internal API for that yet. I can add it if you think it's important.

Comment (carljm) [reply Parser/asdl_c.py:998]: No, I don't think it's important enough for that. This should trivially show up in testing of any hypothetical new generic type (which seems very unlikely in the first place.)

File: Parser/asdl_c.py
Comment (carljm) [reply Parser/asdl_c.py:1056]: Why would the `__new__` call raise deprecation warnings? Aren't the deprecation warnings in `__init__`, not `__new__`? That's why `object.reduce()` uses `__new__`, so that it can bypass whatever is happening in custom `__init__` methods (otherwise lots of custom classes would fail to unpickle by default). So `object.reduce()` is not equivalent to `node = ast.FunctionDef(); node.__dict__.update(...)` -- it's more like `node = ast.FunctionDef.__new__(ast.FunctionDef); node.__dict__.update(...)`, which is a critical difference. So it seems to me like using the default reducer should work fine here, and not require all this extra work. But I could definitely be missing something; there's probably some reason these nodes had a custom reduce() in the first place.

Comment (JelleZijlstra) [reply Parser/asdl_c.py:1056]: I see what you mean. I tried commenting out the `__reduce__` definition, and lots of tests fail with:

```
ERROR: test_pickling (test.test_ast.AST_Tests.test_pickling) (ast=<ast.Module object at 0x102323da0>, protocol=1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jelle/py/cpython/Lib/test/test_ast.py", line 977, in test_pickling
    ast2 = pickle.loads(pickle.dumps(ast, protocol))
                        ~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/jelle/py/cpython/Lib/copyreg.py", line 71, in _reduce_ex
    state = base(self)
            ~~~~^^^^^^
TypeError: AST constructor takes at most 0 positional arguments
```

I don't entirely understand what `_reduce_ex` is doing there, but it picks `ast.AST` as the base, which fails.

Comment (carljm) [reply Parser/asdl_c.py:1056]: Interesting. Well, I might have just nerd-sniped myself into digging into this more later to understand what's happening, but I don't think it should block the PR; your added code to use positional args works fine.

**Review by carljm — APPROVED** [review]
If you are satisfied with the Discourse thread outcome, the code here LGTM! Thanks for this improvement.

File: Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst
Comment (AlexWaygood) [reply Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst:7]: I think maybe this might be worth a short note in "What's new in Python 3.13" as well?

Comment (JelleZijlstra) [reply Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst:7]: Yes, I should also mention it in `ast.rst`.

Comment (JelleZijlstra) [reply Misc/NEWS.d/next/Core and Builtins/2023-06-16-21-29-06.gh-issue-105858.Q7h0EV.rst:7]: Just pushed some docs changes, let me know what you think.

File: Doc/library/ast.rst
Comment (AlexWaygood) [inline Doc/library/ast.rst]:
```suggestion
   Previous versions of Python allowed the creation of AST nodes that were missing
```

File: Doc/library/ast.rst
Comment (AlexWaygood) [inline Doc/library/ast.rst]:
```suggestion
   arguments that were set as attributes of the AST node, even if they did not
```

File: Doc/library/ast.rst
Comment (AlexWaygood) [inline Doc/library/ast.rst]:
```suggestion
.. deprecated-removed:: 3.13 3.15
```

File: Doc/whatsnew/3.13.rst
Comment (AlexWaygood) [inline Doc/whatsnew/3.13.rst]: This all feels a bit terse and hard-to-understand to me right now. I like the wording you put in `ast.rst` more.

Maybe something like this? Could possibly consider a short REPL example as well

```suggestion
* The constructors of node types in the :mod:`ast` module are now stricter
  in the arguments they accept, and have more intuitive behaviour when
  arguments are omitted.

  If an optional field on an AST node is not included as an argument when
  calling that node's constructor method, the field will now be set to
  ``None``. Similarly, if an AST node has a list field and that field is
  omitted when calling the node's constructor method, that field will now be
  set to an empty list. (Previously, in both cases, the attribute would be
  missing on the newly constructed AST node instance.)

  If there are fewer arguments passed to an AST node constructor than there are
  fields on that AST node, a :exc:`DeprecationWarning` will now be emitted.
  Passing too few arguments to the constructor will cause an exception in
  Python 3.15. Similarly, passing a keyword argument that does not map to a
  field on the AST node is now deprecated, and will cause an exception to be
  raised in Python 3.15.
```

Comment (carljm) [reply Doc/whatsnew/3.13.rst]: I suspect we should not use the suggested wording here in the third para "fewer arguments passed than there are fields on the node," as that wording is not really true (it fails to consider optional and list fields.) We could maybe say "fewer arguments passed than there are _required_ fields on the node," but I still think that's less clear than just saying "if no value is provided for a required argument..."

Otherwise (for the first two paras) I find the suggested wording fine, probably a bit clearer, also maybe a bit more verbose than necessary in a few spots (do we really need to say "Similarly, if an AST node has a list field and that field is omitted..." rather than just "Similarly, if a list field is omitted..."?) But I generally find either the PR or suggested-edit text basically fine.

Comment (JelleZijlstra) [reply Doc/whatsnew/3.13.rst]: Thanks both. I pushed a version that is based on Alex's wording, but with the correction Carl brought up and edited for conciseness.

**Comment by JelleZijlstra** [comment]
Since the feedback on Discourse was positive and Carl approved this PR, I'll hit the merge button tomorrow unless new feedback comes up.

File: Doc/whatsnew/3.13.rst
Comment (AlexWaygood) [reply Doc/whatsnew/3.13.rst]: Thanks, the new wording looks great!

**Comment by AlexWaygood** [comment]
> Since the feedback on Discourse was positive and Carl approved this PR, I'll hit the merge button tomorrow unless new feedback comes up.

No objections from me! The only thing that came up on Discourse that I thought was a really good idea and that this PR doesn't already do was to think about possibly prohibiting instantiation of the abstract nodes such as `ast.expr`. But that's a pretty separate idea from the improvements this PR makes, so there's no real reason it needs to be done here.

**Comment by JelleZijlstra** [comment]
Yes, I'll probably create a separate issue for that. One other thing I might want to change is to make the default `repr()` more useful:

```
>>> import ast
>>> ast.Load()
<ast.Load object at 0x103b8ecc0>
>>> ast.Name(id="hello", ctx=ast.Load())
<ast.Name object at 0x103b8d3c0>
```

However, that is also something for a separate issue.

**Comment by JelleZijlstra** [comment]
Clearly something is horribly wrong with the build, I'll take a look soon.

**Comment by JelleZijlstra** [comment]
Reproduces locally with `./python.exe -m test -u cpu test_peg_generator -v`. Now looking into ways to let the extension build by the PEG generator access this type, or find some workaround.

File: Parser/asdl_c.py
Comment (serhiy-storchaka) [inline Parser/asdl_c.py:1092]: There are leaks of `name` here. They will be fixed in #116438.

**Pull request merged**
